go/ngql/types.go

```diff
@@ -330,9 +341,32 @@ func getArgValue(arg interface{}, nomsType *types.Type) (types.Value, error) {
 	return nil, fmt.Errorf("%s is not a subtype of %s", nomsVal.Type().Describe(), nomsType.Describe())
 }
 
-func getCollectionArgs(typ *types.Type, len uint64, args map[string]interface{}, factory iteratorFactory) (iter interface{}, nomsKey, nomsThrough types.Value, count uint64, singleExactMatch bool, err error) {
+type nomsCollection interface {
+	types.Value
+	Len() uint64
+}
+
+func getCollectionArgs(col nomsCollection, args map[string]interface{}, factory iteratorFactory) (iter interface{}, nomsKey, nomsThrough types.Value, count uint64, singleExactMatch bool, err error) {
+	typ := col.Type()
+	len := col.Len()
 	nomsKeyType := typ.Desc.(types.CompoundDesc).ElemTypes[0]
 
+	if keys, ok := args[keysKey]; ok {
+		var nomsValue types.Value
+		nomsValue, err = getArgValue(keys, types.MakeListType(nomsKeyType))
```

Comment:
I made the change but I don't really like it. I understand that creating a types.List and an Iterator for that is not cheap but having to duplicate this code just to optimize this case seems bad. I'll upload my changes and you can look.